Crash during loading with multiple WorkerThreads enabled on Intel(R) Core(TM) i3-2100 CPU @ 3.10GHz

### Observed behaviour
Crash during loading, always in the same place and manner. 

Console log  with gdb session (last lines):
```
Loading [92%]: PostLoad took 0.03ms
Pioneer loading took 6150.41ms
Creating new galaxy generator 'legacy' version 1
Clearing and re-using previous Galaxy object
StarSystemCache: misses: 0, slave hits: 0, master hits: 0
SectorCache: misses: 101, slave hits: 0, master hits: 1
Created shader starfield (address=0x555556ca8e30)
Stars picked from galaxy: 103983
Generating 21017 random stars
/usr/include/c++/12.1.0/bits/stl_vector.h:1123: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = vector3<float>; _Alloc = std::allocator<vector3<float> >; reference = vector3<float>&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.

Thread 1 "pioneer" received signal SIGABRT, Aborted.
0x00007ffff6e8e36c in ?? () from /usr/lib/libc.so.6
```
Gdb backtrace:
```
(gdb) bt
#0  0x00007ffff6e8e36c in ?? () from /usr/lib/libc.so.6
#1  0x00007ffff6e3e838 in raise () from /usr/lib/libc.so.6
#2  0x00007ffff6e28535 in abort () from /usr/lib/libc.so.6
#3  0x00007ffff72d2002 in std::__glibcxx_assert_fail (file=<optimized out>, line=<optimized out>, function=<optimized out>, condition=<optimized out>)
    at /usr/src/debug/gcc/libstdc++-v3/src/c++11/debug.cc:60
#4  0x00005555555f467e in ?? ()
#5  0x00005555556f6b21 in ?? ()
#6  0x00005555556f6f35 in ?? ()
#7  0x00005555556f7fba in ?? ()
#8  0x000055555572b086 in ?? ()
#9  0x000055555572b259 in ?? ()
#10 0x0000555555600ab9 in ?? ()
#11 0x00005555555bdecc in ?? ()
#12 0x00005555555be069 in ?? ()
#13 0x00005555555a8cb6 in ?? ()
#14 0x00007ffff6e29290 in ?? () from /usr/lib/libc.so.6
#15 0x00007ffff6e2934a in __libc_start_main () from /usr/lib/libc.so.6
#16 0x00005555555a8065 in ?? ()
```
I know, it's not very helpful, depsite I added "-D CMAKE_BUILD_TYPE=Debug" and "options=(debug !strip)" to the PKBUILD file. Maybe I should recompile everything again, but I'm afraid I would take long time again. Well, I'll try it, but meanwhile can anybody already look at it?

### Expected behaviour
Normal loading and work, as before about two or three weeks ago.

### Steps to reproduce


__My pioneer version (and OS):__
pioneer-git 20220203.r74.gfeb4169a0-1 (from there: https://aur.archlinux.org/packages/pioneer-git)
OS: Arch Linux, up to date.

EDIT: Impaktor added code formatiting to not reference issues 1-16

I've compiled commit [291a495](https://github.com/pioneerspacesim/pioneer/commit/291a4953b8276f0a2ab55dfbf57cd9041217572c) just before commit "[Web-eWorks/multithread-improvements](https://github.com/pioneerspacesim/pioneer/commit/615f9998ca5605b587de0560432200725b54bd3c)" and I can load the game without problem, but I can't save the game. When I save the gamme appears this error:

/usr/include/c++/12.1.0/bits/stl_vector.h:1142: std::vector<_Tp, _Alloc>::const_reference std::vector<_Tp, _Alloc>::operator[](size_type) const [with _Tp = unsigned int; _Alloc = std::allocator<unsigned int>; const_reference = const unsigned int&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.

And when I'm compiling appears this warnings:
[ 55%] Building CXX object CMakeFiles/pioneer-lib.dir/src/pigui/PiGuiView.cpp.o
In file included from /home/archilolo/builds/pioneer-git/src/pioneer-git/src/lua/LuaMetaType.h:10,
                 from /home/archilolo/builds/pioneer-git/src/pioneer-git/src/pigui/LuaRadar.cpp:7:
/home/archilolo/builds/pioneer-git/src/pioneer-git/src/lua/LuaTable.h:189:37: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
  189 |         class VecIter : public std::iterator<std::input_iterator_tag, Value> {
      |                                     ^~~~~~~~
In file included from /usr/include/c++/12.1.0/bits/stl_algobase.h:65,
                 from /usr/include/c++/12.1.0/list:60,
                 from /usr/include/sigc++-2.0/sigc++/signal.h:7,
                 from /usr/include/sigc++-2.0/sigc++/sigc++.h:123,
                 from /home/archilolo/builds/pioneer-git/src/pioneer-git/src/DeleteEmitter.h:12,
                 from /home/archilolo/builds/pioneer-git/src/pioneer-git/src/lua/LuaObject.h:7,
                 from /home/archilolo/builds/pioneer-git/src/pioneer-git/src/pigui/LuaPiGui.h:7,
                 from /home/archilolo/builds/pioneer-git/src/pioneer-git/src/pigui/LuaRadar.cpp:4:
/usr/include/c++/12.1.0/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator

It seems ArchLinux uses too many new versions of programs and libraries and perhaps pioneer is using an iterator considered deprecated.

> I'm having the same issue. I have an AMD Ryzen 5 2600 (12) @ 3.400GHz CPU and 8 GiB RAM. My memory is not damaged.

Good to know! I wasn't expecting it to be actually related to corrupted memory, but unfortunately system errors of that sort mean there's very little actionable information that we can use from the bug report as nothing about the system can be considered in a consistent, stable state. 

> I'm haven't experienced this issue before. It's the first time. As @pirogronian indicates, using WorkerThreads=1 it works. It seems an issue related to multithreading.

Interestingly, this doesn't actually disable multithreading at all, it just reduces the number of threads involved to a 'core' thread and a 'worker' thread which both still work to fill the vector in question. It's not related to 'cpu type', as a Core-i3 2300 is a *very* different CPU from a Ryzen 5 2600.

...and looking at the code I just realized what the actual problem is. Try adding:

```cpp
stars.pos.resize(NUM_BG_STARS);
stars.color.resize(NUM_BG_STARS);
stars.brightness.resize(NUM_BG_STARS);
```

to `src/Background.cpp` on line 489 (right after `Output("Stars picked from galaxy [...]`) and recompiling. If my guess is correct, this is just a subtle refactoring error that wasn't triggering asserts in my build configuration for some reason.

> /home/archilolo/builds/pioneer-git/src/pioneer-git/src/lua/LuaTable.h:189:37: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]

This is a completely unrelated warning (the code it's warning about is not even in scope when the error is triggered) and is not relevant to the problem at hand. Inheriting from std::iterator will remain deprecated-but-allowed for a very long time.

--- a/src/Background.cpp
+++ b/src/Background.cpp
@@ -487,6 +487,10 @@ namespace Background {
 		num = stars.pos.size();
 		Output("Stars picked from galaxy: %d\n", stars.pos.size());
 
+		stars.pos.resize(NUM_BG_STARS);
+		stars.color.resize(NUM_BG_STARS);
+		stars.brightness.resize(NUM_BG_STARS);
+
 		PROFILE_START_DESC("Generate Random Stars")
 		// fill out the remaining target count with generated points and also fill hyperspace stars
 		Output("Generating %d random stars\n", NUM_BG_STARS - num);

